File: lab/formal-system/concerns/meta-rules/tools/meta_rules_check.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
meta_rules_check —— 把元规则 M1+M2 接入一条**可运行**的检查（支持**递归子关注点**）。

判定"关注点"目录是否符合：
  M1 文件组织=关注点优先/类型其次（含递归）：关注点根有 `README.md`(出入口)；子项是"类型目录 / 根文档 /
      **子关注点**"；子关注点**递归**校验（必须有 README，且内部同样遵守类型词表+命名）。
  M2 命名/类型词表：子目录在**类型词表**内；根文档在白名单或符合命名模式；**文档类**类型内文件命名符合
      （`NN_`/`YYYYMMDD(HHMM)-`/`snake_case`）；代码类类型按各自工程约定豁免。

规则依据：[`../meta-rules/M1-file-organization.md`](../meta-rules/M1-file-organization.md)、
[`../meta-rules/M2-naming-vocabulary.md`](../meta-rules/M2-naming-vocabulary.md)、
配置：[`../meta-rules/meta-rules-config.json`](../meta-rules/meta-rules-config.json)。

用法：
  python3 lab/formal-system/concerns/meta-rules/tools/meta_rules_check.py                    # 默认检查 lab/formal-system
  python3 lab/formal-system/concerns/meta-rules/tools/meta_rules_check.py --root <dir> --config <json>
  python3 lab/formal-system/concerns/meta-rules/tools/meta_rules_check.py --json <path>     # 额外写机器可读结果
  python3 lab/formal-system/concerns/meta-rules/tools/meta_rules_check.py --self-test
退出码：0=全部符合；1=存在违规。
"""
import argparse
import json
import os
import re
import sys
import tempfile

HERE = os.path.dirname(os.path.abspath(__file__))
# 本脚本位于 lab/formal-system/concerns/meta-rules/tools/ 下
DEFAULT_ROOT = os.path.normpath(os.path.join(HERE, "..", "..", ".."))  # lab/formal-system (formal-system)
DEFAULT_CONFIG = os.path.normpath(os.path.join(HERE, "..", "meta-rules-config.json"))


def _clean_cfg(cfg):
    """补齐可选键；sub_configs 若缺省则继承父级命名。"""
    return {
        "root_docs": cfg.get("root_docs", ["README.md"]),
        "type_vocab": cfg.get("type_vocab", []),
        "code_types": cfg.get("code_types", []),
        "empty_ok_types": cfg.get("empty_ok_types", []),
        "doc_naming": cfg.get("doc_naming", {"default": ["^[a-z][a-z0-9_-]*$"]}),
        "skip_dirs": cfg.get("skip_dirs", ["__pycache__", "target", ".git", "node_modules"]),
        "skip_prefixes": cfg.get("skip_prefixes", [".", "_"]),
        "sub_concerns": cfg.get("sub_concerns", []),
        "sub_concerns_dir": cfg.get("sub_concerns_dir"),
        "sub_configs": cfg.get("sub_configs", {}),
    }


def _named(match_regs, name):
    return any(r.match(name) for r in match_regs)


def run_check(root, cfg, prefix=""):
    """递归校验一个关注点；返回 (issues, info) 字符串列表（行已带前缀）。"""
    cfg = _clean_cfg(cfg)
    root = os.path.abspath(root)
    issues, info = [], []

    def mark(ok, text):
        line = f"{prefix}{text}"
        (info if ok else issues).append(line)

    # ---- M1: 出入口 README ----
    if not os.path.isfile(os.path.join(root, "README.md")):
        mark(False, f"[M1] 关注点缺出入口 README.md")
    else:
        mark(True, f"[M1] 出入口 README.md ✔")

    root_docs = set(cfg["root_docs"])
    type_vocab = set(cfg["type_vocab"])
    code_types = set(cfg["code_types"])
    sub_concerns = set(cfg["sub_concerns"])
    sub_concerns_dir = cfg.get("sub_concerns_dir")
    skill_dirs = set(cfg["skip_dirs"])
    default_naming = [re.compile(p) for p in cfg["doc_naming"].get("default", [])]

    def validate_sub(ep, sub_cfg, child_label, child_prefix):
        """校验一个子关注点（代码型 或 递归）。"""
        if sub_cfg.get("code"):
            if os.path.isfile(os.path.join(ep, "README.md")):
                mark(True, f"[M1] 代码子关注点(命名豁免) ✔ {child_label}/")
            else:
                issues.append(f"{prefix}[M1] 代码子关注点缺 README ✗ {child_label}/")
        else:
            sub_issues, sub_info = run_check(ep, sub_cfg, prefix=child_prefix)
            issues.extend(sub_issues)
            info.extend(sub_info)
            if not sub_issues:
                mark(True, f"[M1] 子关注点 ✔ {child_label}/")
            else:
                issues.append(f"{prefix}[M1] 子关注点有违规 ✗ {child_label}/")

    for entry in sorted(os.listdir(root)):
        ep = os.path.join(root, entry)
        if entry in skill_dirs:
            continue
        if os.path.isdir(ep):
            if entry in type_vocab or entry in code_types:
                kind = "代码类(命名豁免)" if entry in code_types else "文档/类型"
                mark(True, f"[M2] 类型目录 ✔ {entry} ({kind})")
            elif sub_concerns_dir and entry == sub_concerns_dir:
                # 子关注点统一目录：其下每个子目录是一个子关注点(允许 README/STRUCTURE 说明文件)
                for sub in sorted(os.listdir(ep)):
                    if sub in ("README.md", "STRUCTURE.md"):
                        continue
                    sub_ep = os.path.join(ep, sub)
                    if os.path.isdir(sub_ep):
                        validate_sub(sub_ep, cfg["sub_configs"].get(sub, {}),
                                     child_label=sub, child_prefix=prefix + entry + "/" + sub + "/")
                    else:
                        mark(False, f"[M1] 子关注点容器内非常规条目 ✗ {entry}/{sub}")
                mark(True, f"[M1] 子关注点统一目录 ✔ {entry}/")
            elif entry in sub_concerns or (
                entry not in type_vocab and os.path.isfile(os.path.join(ep, "README.md"))
            ):
                validate_sub(ep, cfg["sub_configs"].get(entry, {}),
                             child_label=entry, child_prefix=prefix + entry + "/")
            else:
                mark(False, f"[M2] 未知类型目录(不在词表,也非子关注点) ✗ {entry}")
        else:
            if entry in root_docs or entry == "STRUCTURE.md" or _named(default_naming, entry):
                mark(True, f"[M2] 根文档 ✔ {entry}")
            else:
                mark(False, f"[M2] 根目录未声明文件 ✗ {entry}")

    # 子关注点统一目录 / 声明的子关注点须已建立
    if sub_concerns_dir and not os.path.isdir(os.path.join(root, sub_concerns_dir)):
        mark(False, f"[M1] 声明了子关注点统一目录但未建立 ✗ {sub_concerns_dir}/")
    for name in sorted(sub_concerns):
        if not os.path.isdir(os.path.join(root, name)):
            mark(False, f"[M1] 声明了子关注点但目录未建立 ✗ {name}/")

    # ---- M2: 文档类类型目录内命名 ----
    for t in sorted(type_vocab - code_types):
        tpath = os.path.join(root, t)
        if not os.path.isdir(tpath):
            if t in cfg["empty_ok_types"]:
                mark(True, f"[M2] 空类型(允许, 待写) ✔ {t}/")
            continue
        patterns = cfg["doc_naming"].get(t, cfg["doc_naming"].get("default", []))
        regs = [re.compile(p) for p in patterns]
        for name in sorted(os.listdir(tpath)):
            if name in ("README.md", "STRUCTURE.md"):
                continue
            if any(name.startswith(pfx) for pfx in cfg["skip_prefixes"]):
                continue
            if os.path.isdir(os.path.join(tpath, name)):
                mark(True, f"[M2] 子目录(子关注点/再切) · {t}/{name}")
                continue
            if any(r.match(name) for r in regs):
                mark(True, f"[M2] 命名 ✔ {t}/{name}")
            else:
                mark(False, f"[M2] 命名不符合约定 ✗ {t}/{name}  允许:{patterns}")

    return issues, info


def self_test():
    base = tempfile.mkdtemp(prefix="meta-check-")
    # 一级关注点
    open(os.path.join(base, "README.md"), "w").close()
    open(os.path.join(base, "stray.txt"), "w").close()
    os.makedirs(os.path.join(base, "unknown_thing"))
    os.makedirs(os.path.join(base, "docs"))
    open(os.path.join(base, "docs", "My File.md"), "w").close()
    open(os.path.join(base, "docs", "00_ok.md"), "w").close()
    # 递归子关注点：带 README 的最小合法子关注点 → 递归应通过
    os.makedirs(os.path.join(base, "sub_a"))
    open(os.path.join(base, "sub_a", "README.md"), "w").close()

    cfg = {
        "root_docs": ["README.md"],
        "type_vocab": ["docs"],
        "code_types": [],
        "empty_ok_types": [],
        "doc_naming": {"default": ["^\\d{2}_", "^[a-z][a-z0-9_-]*$"]},
        "skip_dirs": [],
        "skip_prefixes": [".", "_"],
        "sub_concerns": ["sub_a"],
    }
    issues, _info = run_check(base, cfg)

    def has(issues_, sub):
        return any(sub in i for i in issues_)

    # 检出三类违规 + 根 README 正常
    ok = (has(issues, "未知类型目录") and has(issues, "根目录未声明文件")
          and has(issues, "命名不符合约定") and not has(issues, "[M1]"))

    # 递归：sub_a 是带 README 的子关注点（命名允许）→ 不应计入违规
    if any("[M1] 子关注点 ✔ sub_a/" in i for i in _info):
        ok = ok and True
    else:
        ok = ok and False

    # 反例：无 README 的子目录（非类型）→ 应按"未知类型目录"被检出
    base2 = tempfile.mkdtemp(prefix="meta-check-")
    open(os.path.join(base2, "README.md"), "w").close()
    os.makedirs(os.path.join(base2, "sub_b"))
    issues2, _ = run_check(base2, cfg)
    ok = ok and has(issues2, "未知类型目录")

    for i in issues:
        print("  捕获: " + i)
    print("self-test:", "PASS" if ok else "FAIL")
    return 0 if ok else 1


def main():
    if "--self-test" in sys.argv:
        return self_test()

    ap = argparse.ArgumentParser(description="元规则 M1+M2 可运行检查(支持递归子关注点)")
    ap.add_argument("--root", default=DEFAULT_ROOT)
    ap.add_argument("--config", default=DEFAULT_CONFIG)
    ap.add_argument("--json", default=None, help="额外写出机器可读结果 JSON")
    args = ap.parse_args()

    with open(args.config, encoding="utf-8") as f:
        cfg = json.load(f)
    issues, info = run_check(args.root, cfg)

    concern = cfg.get("concern", os.path.basename(os.path.abspath(args.root)))
    print(f"元规则检查 M1+M2 · {concern}")
    for line in info:
        print("  ✓ " + line)
    for line in issues:
        print("  ✗ " + line)
    code_hint = ", ".join(sorted(cfg.get("code_types", []))) if cfg.get("code_types") else "—"
    print(f"\n判定: 合规项 {len(info)} · 违规 {len(issues)}   [代码类豁免命名: {code_hint}]")

    if args.json:
        with open(args.json, "w", encoding="utf-8") as f:
            json.dump({
                "generated": __import__("datetime").datetime.now().astimezone().isoformat(),
                "concern": concern,
                "root": os.path.abspath(args.root),
                "compliant": not bool(issues),
                "counts": {"ok": len(info), "violations": len(issues)},
                "items": [{"status": "ok", "text": t} for t in info]
                         + [{"status": "violation", "text": t} for t in issues],
            }, f, ensure_ascii=False, indent=2)
        print(f"结果已写出: {args.json}")
    return 1 if issues else 0


if __name__ == "__main__":
    sys.exit(main())

File: rules/meta/tools/meta_rules_check.ts
#!/usr/bin/env node
// meta_rules_check —— 把元规则 M1+M2 接入一条**可运行**的检查（支持**递归子关注点**）。
// 跨运行时 TypeScript（node/deno/bun 直接运行）。整合自 lab/formal-system/concerns/meta-rules/tools/meta_rules_check.ts。

import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";

const HERE = path.dirname(decodeURIComponent(new URL(import.meta.url).pathname));
// 本脚本位于 rules/meta/tools/ 下
const DEFAULT_ROOT = path.normalize(path.join(HERE, "..", "..")); // rules/
const DEFAULT_CONFIG = path.normalize(path.join(HERE, "..", "meta-rules-config.json"));

function cleanCfg(cfg: any): any {
  return {
    root_docs: cfg.root_docs ?? ["README.md"],
    type_vocab: cfg.type_vocab ?? [],
    code_types: cfg.code_types ?? [],
    empty_ok_types: cfg.empty_ok_types ?? [],
    doc_naming: cfg.doc_naming ?? { default: ["^[a-z][a-z0-9_-]*$"] },
    skip_dirs: cfg.skip_dirs ?? ["__pycache__", "target", ".git", "node_modules"],
    skip_prefixes: cfg.skip_prefixes ?? [".", "_"],
    sub_concerns: cfg.sub_concerns ?? [],
    sub_concerns_dir: cfg.sub_concerns_dir ?? null,
    sub_configs: cfg.sub_configs ?? {},
  };
}

function isFile(p: string): boolean {
  try {
    return fs.statSync(p).isFile();
  } catch {
    return false;
  }
}

function isDir(p: string): boolean {
  try {
    return fs.statSync(p).isDirectory();
  } catch {
    return false;
  }
}

function listdir(p: string): string[] {
  try {
    return fs.readdirSync(p, { encoding: "utf8" });
  } catch {
    return [];
  }
}

function named(matchRegs: RegExp[], name: string): boolean {
  return matchRegs.some((r) => r.test(name));
}

// Python str(list) 风格：['^\\d{2}_', '^[a-z]...']（单引号、转义反斜杠）
function pyRepr(arr: string[]): string {
  return "[" + arr.map((s) => "'" + s.replace(/\\/g, "\\\\").replace(/'/g, "\\'") + "'").join(", ") + "]";
}

function runCheck(rootIn: string, cfgIn: any, prefix = ""): [string[], string[]] {
  const cfg = cleanCfg(cfgIn);
  const root = path.resolve(rootIn);
  const issues: string[] = [];
  const info: string[] = [];

  const mark = (ok: boolean, text: string): void => {
    const line = prefix + text;
    (ok ? info : issues).push(line);
  };

  // ---- M1: 出入口 README ----
  if (!isFile(path.join(root, "README.md"))) {
    mark(false, "[M1] 关注点缺出入口 README.md");
  } else {
    mark(true, "[M1] 出入口 README.md ✔");
  }

  const root_docs = new Set<string>(cfg.root_docs);
  const type_vocab = new Set<string>(cfg.type_vocab);
  const code_types = new Set<string>(cfg.code_types);
  const sub_concerns = new Set<string>(cfg.sub_concerns);
  const sub_concerns_dir = cfg.sub_concerns_dir;
  const skill_dirs = new Set<string>(cfg.skip_dirs);
  const default_naming: RegExp[] = (cfg.doc_naming.default ?? []).map((p: string) => new RegExp(p));

  const validateSub = (ep: string, sub_cfg: any, child_label: string, child_prefix: string): void => {
    if (sub_cfg.code) {
      if (isFile(path.join(ep, "README.md"))) {
        mark(true, `[M1] 代码子关注点(命名豁免) ✔ ${child_label}/`);
      } else {
        issues.push(`${prefix}[M1] 代码子关注点缺 README ✗ ${child_label}/`);
      }
    } else {
      const [sub_issues, sub_info] = runCheck(ep, sub_cfg, child_prefix);
      issues.push(...sub_issues);
      info.push(...sub_info);
      if (sub_issues.length === 0) {
        mark(true, `[M1] 子关注点 ✔ ${child_label}/`);
      } else {
        issues.push(`${prefix}[M1] 子关注点有违规 ✗ ${child_label}/`);
      }
    }
  };

  const entries = listdir(root).sort();
  for (const entry of entries) {
    const ep = path.join(root, entry);
    if (skill_dirs.has(entry)) continue;
    if (isDir(ep)) {
      if (type_vocab.has(entry) || code_types.has(entry)) {
        const kind = code_types.has(entry) ? "代码类(命名豁免)" : "文档/类型";
        mark(true, `[M2] 类型目录 ✔ ${entry} (${kind})`);
      } else if (sub_concerns_dir && entry === sub_concerns_dir) {
        // 子关注点统一目录：其下每个子目录是一个子关注点(允许 README/STRUCTURE 说明文件)
        const subs = listdir(ep).sort();
        for (const sub of subs) {
          if (sub === "README.md" || sub === "STRUCTURE.md") continue;
          const sub_ep = path.join(ep, sub);
          if (isDir(sub_ep)) {
            validateSub(sub_ep, cfg.sub_configs[sub] ?? {}, sub, prefix + entry + "/" + sub + "/");
          } else {
            mark(false, `[M1] 子关注点容器内非常规条目 ✗ ${entry}/${sub}`);
          }
        }
        mark(true, `[M1] 子关注点统一目录 ✔ ${entry}/`);
      } else if (
        sub_concerns.has(entry) ||
        (!type_vocab.has(entry) && isFile(path.join(ep, "README.md")))
      ) {
        validateSub(ep, cfg.sub_configs[entry] ?? {}, entry, prefix + entry + "/");
      } else {
        mark(false, `[M2] 未知类型目录(不在词表,也非子关注点) ✗ ${entry}`);
      }
    } else {
      if (root_docs.has(entry) || entry === "STRUCTURE.md" || named(default_naming, entry)) {
        mark(true, `[M2] 根文档 ✔ ${entry}`);
      } else {
        mark(false, `[M2] 根目录未声明文件 ✗ ${entry}`);
      }
    }
  }

  // 子关注点统一目录 / 声明的子关注点须已建立
  if (sub_concerns_dir && !isDir(path.join(root, sub_concerns_dir))) {
    mark(false, `[M1] 声明了子关注点统一目录但未建立 ✗ ${sub_concerns_dir}/`);
  }
  for (const name of Array.from(sub_concerns).sort()) {
    if (!isDir(path.join(root, name))) {
      mark(false, `[M1] 声明了子关注点但目录未建立 ✗ ${name}/`);
    }
  }

  // ---- M2: 文档类类型目录内命名 ----
  const docTypes = Array.from(new Set<string>(cfg.type_vocab as string[]))
    .filter((t) => !code_types.has(t))
    .sort();
  for (const t of docTypes) {
    const tpath = path.join(root, t);
    if (!isDir(tpath)) {
      if (cfg.empty_ok_types.includes(t)) {
        mark(true, `[M2] 空类型(允许, 待写) ✔ ${t}/`);
      }
      continue;
    }
    const patterns = cfg.doc_naming[t] ?? cfg.doc_naming.default ?? [];
    const regs: RegExp[] = patterns.map((p: string) => new RegExp(p));
    const names = listdir(tpath).sort();
    for (const name of names) {
      if (name === "README.md" || name === "STRUCTURE.md") continue;
      if (cfg.skip_prefixes.some((pfx: string) => name.startsWith(pfx))) continue;
      if (isDir(path.join(tpath, name))) {
        mark(true, `[M2] 子目录(子关注点/再切) · ${t}/${name}`);
        continue;
      }
      if (named(regs, name)) {
        mark(true, `[M2] 命名 ✔ ${t}/${name}`);
      } else {
        mark(false, `[M2] 命名不符合约定 ✗ ${t}/${name}  允许:${pyRepr(patterns)}`);
      }
    }
  }

  return [issues, info];
}

function selfTest(): number {
  const base = fs.mkdtempSync(path.join(os.tmpdir(), "meta-check-"));
  // 一级关注点
  fs.writeFileSync(path.join(base, "README.md"), "");
  fs.writeFileSync(path.join(base, "stray.txt"), "");
  fs.mkdirSync(path.join(base, "unknown_thing"));
  fs.mkdirSync(path.join(base, "docs"));
  fs.writeFileSync(path.join(base, "docs", "My File.md"), "");
  fs.writeFileSync(path.join(base, "docs", "00_ok.md"), "");
  // 递归子关注点：带 README 的最小合法子关注点 → 递归应通过
  fs.mkdirSync(path.join(base, "sub_a"));
  fs.writeFileSync(path.join(base, "sub_a", "README.md"), "");

  const cfg = {
    root_docs: ["README.md"],
    type_vocab: ["docs"],
    code_types: [],
    empty_ok_types: [],
    doc_naming: { default: ["^\\d{2}_", "^[a-z][a-z0-9_-]*$"] },
    skip_dirs: [],
    skip_prefixes: [".", "_"],
    sub_concerns: ["sub_a"],
  };
  const [issues, _info] = runCheck(base, cfg);

  const has = (list: string[], sub: string): boolean => list.some((i) => i.includes(sub));

  // 检出三类违规 + 根 README 正常
  let ok =
    has(issues, "未知类型目录") &&
    has(issues, "根目录未声明文件") &&
    has(issues, "命名不符合约定") &&
    !has(issues, "[M1]");

  // 递归：sub_a 是带 README 的子关注点（命名允许）→ 不应计入违规
  ok = ok && _info.some((i) => i.includes("[M1] 子关注点 ✔ sub_a/"));

  // 反例：无 README 的子目录（非类型）→ 应按"未知类型目录"被检出
  const base2 = fs.mkdtempSync(path.join(os.tmpdir(), "meta-check-"));
  fs.writeFileSync(path.join(base2, "README.md"), "");
  fs.mkdirSync(path.join(base2, "sub_b"));
  const [issues2] = runCheck(base2, cfg);
  ok = ok && has(issues2, "未知类型目录");

  for (const i of issues) console.log("  捕获: " + i);
  console.log("self-test:", ok ? "PASS" : "FAIL");
  return ok ? 0 : 1;
}

function argGet(argv: string[], name: string): string | null {
  const flag = "--" + name;
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === flag) {
      if (i + 1 < argv.length) return argv[i + 1];
      return null;
    }
    if (argv[i].startsWith(flag + "=")) return argv[i].slice(flag.length + 1);
  }
  return null;
}

function main(): number {
  if (process.argv.includes("--self-test")) return selfTest();

  const root = argGet(process.argv, "root") ?? DEFAULT_ROOT;
  const config = argGet(process.argv, "config") ?? DEFAULT_CONFIG;

  let cfg: any;
  try {
    cfg = JSON.parse(fs.readFileSync(config, "utf8"));
  } catch (e) {
    console.error("配置文件读取失败: " + config);
    return 1;
  }
  const [issues, info] = runCheck(root, cfg);

  const concern = cfg.concern ?? path.basename(path.resolve(root));
  console.log(`元规则检查 M1+M2 · ${concern}`);
  for (const line of info) console.log("  ✓ " + line);
  for (const line of issues) console.log("  ✗ " + line);
  const codeTypes: string[] = cfg.code_types ?? [];
  const codeHint = codeTypes.length ? codeTypes.slice().sort().join(", ") : "—";
  console.log(`\n判定: 合规项 ${info.length} · 违规 ${issues.length}   [代码类豁免命名: ${codeHint}]`);

  const jsonPath = argGet(process.argv, "json");
  if (jsonPath) {
    const payload = {
      generated: new Date().toISOString(),
      concern,
      root: path.resolve(root),
      compliant: issues.length === 0,
      counts: { ok: info.length, violations: issues.length },
      items: info.map((t) => ({ status: "ok", text: t })).concat(issues.map((t) => ({ status: "violation", text: t }))),
    };
    try {
      fs.writeFileSync(jsonPath, JSON.stringify(payload, null, 2));
      console.log("结果已写出: " + jsonPath);
    } catch (e) {
      console.error("结果写入失败: " + jsonPath);
      return 1;
    }
  }
  return issues.length ? 1 : 0;
}

process.exitCode = main();
